segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -0,0 +1,510 @@
+                data.setVersion(committedVersion);
+                toAdd.put(key, data);
+            }
+            bufferedTxnData.putAll(toAdd);
+        }
+
+        //  Step 5 : evict if required.
+        if (bufferedTxnData.size() > maxEntriesInTxnBuffer) {
+            bufferedTxnData.entrySet().removeIf(entry -> entry.getValue().isPersisted() && !entry.getValue().isPinned());
+        }
+
+        //  Step 6: finally clear
+        txnData.clear();
+    }
+
+    /**
+     * Aborts given transaction.
+     *
+     * @param txn transaction to abort.
+     * @throws StorageMetadataException If there are any errors.
+     */
+    public void abort(MetadataTransaction txn) throws StorageMetadataException  {
+        // Do nothing
+    }
+
+    /**
+     * Retrieves the metadata for given key.
+     *
+     * @param txn Transaction.
+     * @param key key to use to retrieve metadata.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     * @return Metadata for given key. Null if key was not found.
+     */
+    @Override
+    public StorageMetadata get(MetadataTransaction txn, String key) throws StorageMetadataException  {
+        Preconditions.checkArgument(null != txn);
+        TransactionData dataFromBuffer = null;
+        if (null == key) {
+            return null;
+        }
+
+        Map<String, TransactionData> txnData = txn.getData();
+        TransactionData data = txnData.get(key);
+
+        // Search in the buffer.
+        if (null == data) {
+            synchronized (lock) {
+                dataFromBuffer = bufferedTxnData.get(key);
+            }
+            // If we did not find in buffer then load it from store
+            if (null == dataFromBuffer) {
+                // NOTE: This call to read MUST be outside the lock, it is most likely cause re-entry.
+                loadFromStore(key);
+                dataFromBuffer = bufferedTxnData.get(key);
+                Preconditions.checkState(null != dataFromBuffer);
+            }
+
+            if (null != dataFromBuffer && null != dataFromBuffer.getValue()) {
+                // Make copy.
+                data = dataFromBuffer.toBuilder()
+                        .key(key)
+                        .value(dataFromBuffer.getValue().copy())
+                        .build();
+                txnData.put(key, data);
+            }
+        }
+
+        if (data != null) {
+            return data.getValue();
+        }
+        return null;
+    }
+
+    /**
+     * Loads value from store
+     * @param key Key to load
+     * @return Value if found null otherwise.
+     * @throws StorageMetadataException Any exceptions.
+     */
+    @VisibleForTesting
+    private TransactionData loadFromStore(String key) throws StorageMetadataException {
+        // NOTE: This call to read MUST be outside the lock, it is most likely cause re-entry.
+        log.debug("Loading key from the store key = {}", key);
```

Comment:
Too much logging here. Use only one debug statement. IF you want more, change one to trace.

Reply:
fixed.

Reply:
made it trace